src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,842 @@
+                }
+            }
+
+            _inObject = _bitStack.Pop();
+        }
+
+        private void WriteEndIndented(byte token)
+        {
+            // Do not format/indent empty JSON object/array.
+            if ((_tokenType == JsonTokenType.StartObject && token == JsonConstants.CloseBrace)
+                || (_tokenType == JsonTokenType.StartArray && token == JsonConstants.CloseBracket))
+            {
+                WriteEndMinimized(token);
+            }
+            else
+            {
+                int idx = 0;
+                WriteNewLine(ref idx);
+
+                int indent = Indentation;
+                while (true)
+                {
+                    bool result = JsonWriterHelper.TryWriteIndentation(_buffer.Slice(idx), indent, out int bytesWritten);
+                    idx += bytesWritten;
+                    if (result)
+                    {
+                        break;
+                    }
+                    indent -= bytesWritten;
+                    AdvanceAndGrow(idx);
+                    idx = 0;
+                }
+
+                while (_buffer.Length <= idx)
+                {
+                    AdvanceAndGrow(idx);
+                    idx = 0;
+                }
+                _buffer[idx++] = token;
+
+                Advance(idx);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNewLine(ref int idx)
+        {
+            // Write '\r\n' OR '\n', depending on OS
+            if (JsonWriterHelper.s_newLineUtf8.Length == 2)
+            {
+                while (_buffer.Length <= idx)
+                {
+                    AdvanceAndGrow(idx);
+                    idx = 0;
+                }
+                _buffer[idx++] = JsonConstants.CarriageReturn;
+            }
+
+            while (_buffer.Length <= idx)
+            {
+                AdvanceAndGrow(idx);
+                idx = 0;
+            }
+            _buffer[idx++] = JsonConstants.LineFeed;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void UpdateBitStackOnStart(byte token)
+        {
+            if (token == JsonConstants.OpenBracket)
+            {
+                _bitStack.PushFalse();
+                _inObject = false;
+            }
+            else
+            {
+                Debug.Assert(token == JsonConstants.OpenBrace);
+                _bitStack.PushTrue();
+                _inObject = true;
+            }
+        }
+
+        private void GrowAndEnsure()
+        {
+            Flush();
+            int previousSpanLength = _buffer.Length;
+            _buffer = _output.GetSpan(DefaultGrowthSize);
+            if (_buffer.Length <= previousSpanLength)
+            {
+                _buffer = _output.GetSpan(DefaultGrowthSize);
+                if (_buffer.Length <= previousSpanLength)
```

Comment:
Why are we doing the exact same thing twice?  And if our goal is to get something at least previousSpanLength in length, why not request Math.Max(previousSpanLength, DefaultGrowthSize)?

Reply:
This is partly due to the ambiguous nature of the IBufferWriter contract which we should resolve and document clearly. Since the GetSpan() call accepts a size hint, there is no guarantee that the implementation returns the size you asked for (it won't return an empty buffer).

The reason for the double check is to try to account for that. For example, if someone provides an IBufferWriter that contains empty segments (at most 1), then even though the first request would fail, the second should succeed. Similarly, it could first return the leftover bytes in your current segment instead which may not be enough.

Admittedly, it isn't great. I will remove it and do a single call to keep things simple and clear, and reject strange implementations of the interface.

> And if our goal is to get something at least previousSpanLength in length

The goal is to get a buffer that is larger than before since we know that the current sized buffer is too small.

Reply:
cc @pakrym, @KrzysztofCwalina 

Can you please confirm that it is fine for us to throw if a **single** call to IBufferWriter.GetSpan() does not return a larger buffer? In this case, we have a buffer that is too small and we ask for a buffer of length 4k. If the buffer we get back has not grown (at least by 1 byte), can we safely throw here? Is there a useful scenario where asking for a 4k sized buffer again would succeed?

Reply:
I don't think repeated calls to `IBufferWriter` with the same value are required. 

In general, there are 2 kinds of situation when GetSpan gets called:

1. There is a requirement for continuous uninterrupted space. Example of this would be to serialize an `int`.
2. We just need some space to write data that might be chunked into multiple memory blocks. Example of this is serializing a string.

For case 1 calling `GetSpan` with the required size and verifying that returned memory segment is at least requested size is enough.
For case 2 we should call `GetSpan` without arguments and use whatever space was returned.